### PR TITLE
Add a `/health` endpoint to the `server`

### PR DESCRIPTION
**Pull request by ibehnam**
This PR adds a `/health` endpoint to the `server` example. It works as follows:

- While the model is being loaded, querying `/health` returns `"{"status": "loading model"}`.
- After the model is successfully loaded, querying `/health` returns `{"status": "ok"}`.
- If any errors occurs, `/health` returns `{"status": "error"}`.

I re-ordered the code about the two threads of the server so that the server starts first, but only `/health` is available. In the other thread, the model is being loaded. If the model successfully loads, the rest of the endpoints (`/completion/`, etc.) become available as well.

## Why add a `/health` endpoint?

Because loading large models (especially with `--no-mmap` flag) takes a long time. The server should not accept completions/embedding/etc. requests during this time. The user can query `/health` to make sure the model is fully loaded before they send their requests.

## Example

While the model is being loaded:

```
curl --request GET --url http://127.0.0.1:8080/health
>> {"status": "loading model"}
```

After model is fully loaded:

```
curl --request GET --url http://127.0.0.1:8080/health
>> {"status": "ok"}
```

### CODE REVIEW

**Comment by ibehnam** [comment]
@ggerganov I believe this is related to https://github.com/ggerganov/llama.cpp/issues/4216.

File: examples/server/server.cpp [hunk 1]
```diff
@@ -2790,15 +2799,118 @@ int main(int argc, char **argv)
...
+    httplib::Server svr;
+
+    ServerState server_state = LOADING_MODEL;
+
+    svr.set_default_headers({{"Server", "llama.cpp"},
+                             {"Access-Control-Allow-Origin", "*"},
+                             {"Access-Control-Allow-Headers", "content-type"}});
+
+    svr.Get("/health", [&](const httplib::Request&, httplib::Response& res) {
+        switch(server_state) {
```
Comment (ggerganov) [inline examples/server/server.cpp:2811]: There is a data race on `server_state`. Although not an issue in reality, the thread sanitizer is not going to like this, so lets make it atomic

File: examples/server/server.cpp [hunk 2]
```diff
@@ -2790,15 +2799,118 @@ int main(int argc, char **argv)
...
+                if (!svr.listen_after_bind())
+                {
+                    server_state = ERROR;
+                    return 1;
+                }
+
+                return 0;
+            });
+
+
```
Comment (ggerganov) [inline examples/server/server.cpp:2904]: Single empty line is enough - fix the other places too

**Comment by ibehnam** [comment]
@ggerganov Okay, I made those changes and made `ServerState` atomic. Tested it on my Mac and works.

**Comment by m18coppola** [comment]
This commit seems to have broken some of the logging. The `--api-key` parameter information no longer prints to stdout/stderr/logfile.

**Comment by ibehnam** [comment]
> This commit seems to have broken some of the logging. The `--api-key` parameter information no longer prints to stdout/stderr/logfile.

It does:

```
{"timestamp":1704924424,"level":"INFO","function":"main","line":2888,"message":"HTTP server listening","api_key":"api_key: ****6789","port":"8080","hostname":"127.0.0.1"}
```

It's just that the server starts before loading the LLM, so that information is printed at the top instead of in the end.

**Comment by m18coppola** [comment]
Oh geez... I should've been paying more attention. Thanks!

**Comment by ibehnam** [comment]
No worries, you got it!

**Pull request merged**